### PR TITLE
Simple benchmarker added.

### PR DESCRIPTION
**Pull request by dotPiano**
Simple benchmarker for measuring the execution time:
- Between any two code lines, with BENCHMARK_START(name); BENCHMARK_STOP(name);
- Of a whole scope (e.g. a function), with BENCHMARK_BLOCK(name)

Mean and standard deviation are computed for multiple measurements with the same name.

### CODE REVIEW

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#endif
+
+/// \brief Benchmark helper class. Allows collecting execution times
+/// of parts of the code and computing their statistics.
+/// \note This class is thread-safe.
+class Benchmarker {
+
+ public:
+  /// \brief The clock used for timing operations
+  typedef std::chrono::high_resolution_clock clock;
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: We usually have the definitions starting with a capital letter eg `Clock`

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+/// \brief A timer that measures the time elapsed between its creation
+/// and destruction. This is useful for timing whole functions or any
+/// scoped block of code. The measurement is automatically committed to
+/// the benchmarker.
+class ScopedTimer {
+
+ public:
+  /// \brief The clock used for timing operations
+  typedef std::chrono::high_resolution_clock clock;
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: `Clock`

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  /// \brief Default constructor. Starts the timer.
+  ScopedTimer(const std::string name);
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: const ref

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+};
+
+/// \brief A timer that measures the time elapsed between its creation
+/// and destruction. This is useful for timing whole functions or any
+/// scoped block of code. The measurement is automatically committed to
+/// the benchmarker.
+class ScopedTimer {
+
+ public:
+  /// \brief The clock used for timing operations
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: Always put ending dot.

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  typedef std::chrono::high_resolution_clock clock;
+
+  /// \brief Add a time measurement for a region of code with a specific name.
+  static void addMeasurement(
+      const std::string& name,
+      const std::chrono::time_point<clock>& start,
+      const std::chrono::time_point<clock>& end);
+
+  /// \brief Print the statistics to the specified file
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: ending . after comment

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  static void saveStatistics(const std::string& file_name);
+
+  /// \brief Print the statistics to the logger
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: ending . after comment

File: laser_slam/include/laser_slam/benchmarker.hpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+class ScopedTimer {
+
+ public:
+  /// \brief The clock used for timing operations
+  typedef std::chrono::high_resolution_clock clock;
+
+  /// \brief Default constructor. Starts the timer.
+  ScopedTimer(const std::string name);
+
+  /// \brief Destructor. Stops the timer and commit the result to the
```
Comment (rdube) [inline laser_slam/include/laser_slam/benchmarker.hpp]: commits

File: laser_slam/src/benchmarker.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+#include <iomanip>
+#include <iostream>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <glog/logging.h>
+
+#include "laser_slam/benchmarker.hpp"
```
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: Put that include at the top eg: https://github.com/ethz-asl/segmatch/blob/master/segmatch/src/segmatch.cpp#L1

File: laser_slam/src/benchmarker.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+namespace laser_slam {
+
+void Benchmarker::addMeasurement(
+    const std::string& name,
+    const std::chrono::time_point<clock>& start,
+    const std::chrono::time_point<clock>& end) {
+
+  constexpr double mus_to_ms = 1.0 / 1000.0;
```
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: does it matter having it as a constexpr member variable or is it computed only once anyway?

File: laser_slam/src/benchmarker.cpp [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  std::ofstream out_file;
+  out_file.open (file_name);
+
+  if (out_file.is_open()) {
+
+    out_file << "Benchmark results:" << std::endl;
+    out_file << " " << std::endl
+        << ALIGN_MODIFIERS << "Name: " << "Mean (SD)" << std::endl;
+
+    for (auto& codeBlock : statistics_) {
```
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: `code_block`

File: laser_slam/src/benchmarker.cpp [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+void Benchmarker::logStatistics() {
+
+  LOG(INFO) << "";
+  LOG(INFO) << "Benchmark results:";
+  LOG(INFO) << " "
+      << ALIGN_MODIFIERS << "Name: " << "Mean (SD)";
+
+  for (auto& codeBlock : statistics_) {
```
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: `code_block`

File: laser_slam/src/benchmarker.cpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+  statistics_[name].addMeasurement(milliseconds);
+}
+
+#define ALIGN_MODIFIERS std::setw(40) << std::setfill(' ') << std::left
+#define FLOAT_MODIFIERS std::fixed << std::setprecision(2)
+
+void Benchmarker::saveStatistics(const std::string& file_name) {
+
+  std::ofstream out_file;
+  out_file.open (file_name);
```
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: rm white space after open

File: laser_slam/src/benchmarker.cpp (hunk 10 above)
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: could be a const ref right?

File: laser_slam/src/benchmarker.cpp [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  return sum_ / static_cast<double>(measurements_count_);
+}
+
+double Benchmarker::MeasurementStatistics_::getStandardDeviation() const {
+  return sqrt(
+      sum_of_squares_ / static_cast<double>(measurements_count_) -
+      pow(sum_ / static_cast<double>(measurements_count_), 2.0));
+}
+
+ScopedTimer::ScopedTimer(const std::string name)
```
Comment (rdube) [inline laser_slam/src/benchmarker.cpp]: const ref

File: laser_slam/src/benchmarker.cpp (hunk 9 above)
Comment (dotPiano) [inline laser_slam/src/benchmarker.cpp]: Three different approaches leading to the same (optimal) result: https://godbolt.org/g/LqsSmi
I personally prefer this version in this case for readability :)

File: laser_slam/src/benchmarker.cpp (hunk 10 above)
Comment (dotPiano) [inline laser_slam/src/benchmarker.cpp]: Of course, thanks :)

File: laser_slam/include/laser_slam/benchmarker.hpp (hunk 2 above)
Comment (dotPiano) [inline laser_slam/include/laser_slam/benchmarker.hpp]: Also this should use Benchmarker::Clock, I'll fix!

File: laser_slam/src/benchmarker.cpp (hunk 12 above)
Comment (dotPiano) [inline laser_slam/src/benchmarker.cpp]: Copy-paste from cppreference spotted... :innocent:

**Pull request merged**